File: DataStruct/LeetCode/547.省份数量.go
package LeetCode

import "fmt"

// DSU Disjoint_Set_Union
type DSU struct {
	pa []int
}

func NewDSU(size int) *DSU {
	pa := make([]int, size)
	for i := 0; i < size; i++ {
		pa[i] = i
	}
	return &DSU{pa: pa}
}

// Find 给出 x 所属的集合
func (d *DSU) Find(x int) int {
	// 若 x 的父节点指向自身, 即 x 为当前集合的代表元素
	if d.pa[x] == x {
		return x
	}
	return d.Find(d.pa[x])
}

// Union 把 x 所属集合并入 y 所属集合
func (d *DSU) Union(x, y int) {
	d.pa[d.Find(x)] = d.Find(y)
}

func findCircleNum(isConnected [][]int) int {
	dsu := NewDSU(len(isConnected))
	for a, cities := range isConnected {
		for b, city := range cities {
			if city != 0 && a != b {
				fmt.Println(a, "\t->\t", b)
				dsu.Union(a, b)
			}
		}
	}
	set := make(map[int]struct{})
	for _, v := range dsu.pa {
		set[v] = struct{}{}
	}
	return len(set)
}

File: DataStruct/LeetCode/547.省份数量.py
from typing import List


class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        num = len(isConnected)
        unionFind = UnionFastFind(num)
        for city in range(num):
            for connnectedCity in range(num):
                if isConnected[city][connnectedCity] != 0:
                    unionFind.union(city, connnectedCity)
        return len(set(unionFind.ids))


class FastUnionFind:
    def __init__(self, n):
        self.fa = [i for i in range(n)]

    def find(self, x):
        while self.fa[x] != x:
            x = self.fa[x]
        return x

    def union(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)

        if root_x == root_y:
            return False
        self.fa[root_x] = self.fa[root_y]


class UnionFastFind:
    def __init__(self, n):
        self.ids = [i for i in range(n)]

    def find(self, x):
        return self.ids[x]

    def union(self, x, y) -> bool:
        """
        将 y 集合合并到 x 集合中
        """
        x_id = self.find(x)
        y_id = self.find(y)

        if x_id == y_id: return False
        for i in range(len(self.ids)):
            if self.find(i) == y_id:
                self.ids[i] = x_id
        return True

    def isConnected(self, x, y):
        return self.find(x) == self.find(y)


print(Solution().findCircleNum([[1, 1, 0], [1, 1, 0], [0, 0, 1]]))
print(Solution().findCircleNum([[1, 0, 0], [0, 1, 0], [0, 0, 1]]))
